File: qt/src/core/ImageProcessor.cc
#include "ImageProcessor.hh"

#include <QColor>
#include <QPainter>

#include "utils/Constants.hh"

namespace timelapse {

ImageProcessor::ImageProcessor(QObject* parent)
   : QObject(parent) {}

auto ImageProcessor::resizeWithLetterbox(QImage const& source,
                                         int32_t targetWidth,
                                         int32_t targetHeight)
   -> std::expected<QImage, QString> {
   if (source.isNull()) {
      return std::unexpected("Source image is null");
   }

   // Create a black background image at target size
   QImage result(targetWidth, targetHeight, QImage::Format_RGB32);
   result.fill(Qt::black);

   // Calculate scaling to fit within target while maintaining aspect ratio
   double sourceRatio = static_cast<double>(source.width()) / source.height();
   double targetRatio = static_cast<double>(targetWidth) / targetHeight;

   int32_t scaledWidth;
   int32_t scaledHeight;

   if (sourceRatio > targetRatio) {
      // Source is wider - fit to width
      scaledWidth = targetWidth;
      scaledHeight = static_cast<int32_t>(targetWidth / sourceRatio);
   } else {
      // Source is taller - fit to height
      scaledHeight = targetHeight;
      scaledWidth = static_cast<int32_t>(targetHeight * sourceRatio);
   }

   // Scale the source image
   QImage scaled = source.scaled(scaledWidth, scaledHeight,
                                 Qt::KeepAspectRatio,
                                 Qt::SmoothTransformation);

   // Calculate position to center the scaled image
   int32_t x = (targetWidth - scaledWidth) / 2;
   int32_t y = (targetHeight - scaledHeight) / 2;

   // Draw the scaled image onto the black background
   QPainter painter(&result);
   painter.drawImage(x, y, scaled);
   painter.end();

   return result;
}

auto ImageProcessor::isImageBlack(QImage const& image, double threshold) const -> bool {
   double brightness = calculateMeanBrightness(image);
   return brightness < threshold;
}

auto ImageProcessor::saveImage(QImage const& image, QString const& path)
   -> std::expected<void, QString> {
   if (image.isNull()) {
      return std::unexpected("Cannot save null image");
   }

   if (!image.save(path, "PNG")) {
      return std::unexpected("Failed to save image to: " + path);
   }

   return {};
}

auto ImageProcessor::calculateMeanBrightness(QImage const& image) const -> double {
   if (image.isNull()) {
      return 0.0;
   }

   // Convert to format that allows easy pixel access
   QImage converted = image.convertToFormat(QImage::Format_RGB32);

   double totalBrightness = 0.0;
   int32_t sampleCount = 0;

   // Sample pixels at intervals for performance
   for (int32_t y = 0; y < converted.height(); y += kSampleStep) {
      for (int32_t x = 0; x < converted.width(); x += kSampleStep) {
         QColor pixel(converted.pixel(x, y));

         // Calculate luminance using standard weights
         double luminance = 0.299 * pixel.red() +
                           0.587 * pixel.green() +
                           0.114 * pixel.blue();

         totalBrightness += luminance / 255.0;
         sampleCount++;
      }
   }

   if (sampleCount == 0) {
      return 0.0;
   }

   return totalBrightness / sampleCount;
}

}  // namespace timelapse

File: rust/src/core/image_processor.rs
use image::{imageops, GenericImageView, ImageBuffer, Rgba, RgbaImage};
use std::path::Path;

use crate::core::error::{Error, ErrorCode, Result};
use crate::utils::constants::SAMPLE_STEP;

pub struct ImageProcessor;

impl ImageProcessor {
   pub fn new() -> Self {
      Self
   }

   /// Resize an image with letterboxing to maintain aspect ratio
   pub fn resize_with_letterbox(
      &self,
      source: &RgbaImage,
      target_width: u32,
      target_height: u32,
   ) -> Result<RgbaImage> {
      if source.width() == 0 || source.height() == 0 {
         return Err(Error::new(ErrorCode::ImageProcessingFailed, "Source image is empty"));
      }

      // Create black background
      let mut result: RgbaImage = ImageBuffer::from_pixel(
         target_width,
         target_height,
         Rgba([0, 0, 0, 255]),
      );

      // Calculate scaling
      let source_ratio = source.width() as f64 / source.height() as f64;
      let target_ratio = target_width as f64 / target_height as f64;

      let (scaled_width, scaled_height) = if source_ratio > target_ratio {
         // Source is wider - fit to width
         let w = target_width;
         let h = (target_width as f64 / source_ratio) as u32;
         (w, h)
      } else {
         // Source is taller - fit to height
         let h = target_height;
         let w = (target_height as f64 * source_ratio) as u32;
         (w, h)
      };

      // Scale the source image
      let scaled = imageops::resize(
         source,
         scaled_width,
         scaled_height,
         imageops::FilterType::Lanczos3,
      );

      // Calculate position to center
      let x = (target_width - scaled_width) / 2;
      let y = (target_height - scaled_height) / 2;

      // Copy scaled image onto result
      imageops::overlay(&mut result, &scaled, x as i64, y as i64);

      Ok(result)
   }

   /// Check if an image is mostly black
   pub fn is_image_black(&self, image: &RgbaImage, threshold: f64) -> bool {
      let brightness = self.calculate_mean_brightness(image);
      brightness < threshold
   }

   /// Calculate the mean brightness of an image by sampling pixels
   fn calculate_mean_brightness(&self, image: &RgbaImage) -> f64 {
      if image.width() == 0 || image.height() == 0 {
         return 0.0;
      }

      let mut total_brightness = 0.0;
      let mut sample_count = 0u64;

      for y in (0..image.height()).step_by(SAMPLE_STEP as usize) {
         for x in (0..image.width()).step_by(SAMPLE_STEP as usize) {
            let pixel = image.get_pixel(x, y);

            // Calculate luminance using standard weights
            let luminance = 0.299 * pixel[0] as f64
               + 0.587 * pixel[1] as f64
               + 0.114 * pixel[2] as f64;

            total_brightness += luminance / 255.0;
            sample_count += 1;
         }
      }

      if sample_count == 0 {
         return 0.0;
      }

      total_brightness / sample_count as f64
   }

   /// Save an image to a file as PNG
   pub fn save_image(&self, image: &RgbaImage, path: &Path) -> Result<()> {
      image.save(path).map_err(|e| {
         Error::new(
            ErrorCode::FileWriteError,
            format!("Failed to save image: {}", e),
         )
      })
   }
}

impl Default for ImageProcessor {
   fn default() -> Self {
      Self::new()
   }
}
